src/include/snapper/dialogs.rb

```diff
@@ -47,198 +44,409 @@ def initialize_snapper_dialogs(include_target)
+      end
     end
 
-
     # Popup for modification of existing snapshot
     # @return true if new snapshot was created
     def ModifySnapshotPopup(snapshot)
-      snapshot = deep_copy(snapshot)
       modified = false
-      num = Ops.get_integer(snapshot, "num", 0)
-      pre_num = Ops.get_integer(snapshot, "pre_num", num)
-      type = Ops.get_symbol(snapshot, "type", :none)
 
-      pre_index = Ops.get(Snapper.id2index, pre_num, 0)
-      pre_snapshot = Ops.get(Snapper.snapshots, pre_index, {})
+      open_modify_dialog(modify_content(snapshot))
 
-      snapshot_term = lambda do |prefix, data|
-        data = deep_copy(data)
-        HBox(
-          HSpacing(),
-          Frame(
-            "",
-            HBox(
-              HSpacing(0.4),
-              VBox(
-                # text entry label
-                InputField(
-                  Id(Ops.add(prefix, "description")),
-                  Opt(:hstretch),
-                  _("Description"),
-                  Ops.get_string(data, "description", "")
-                ),
-                # text entry label
-                InputField(
-                  Id(Ops.add(prefix, "userdata")),
-                  Opt(:hstretch),
-                  _("User data"),
-                  Snapper.userdata_to_string(data["userdata"])
-                ),
-                Left(
-                  ComboBox(
-                    Id(Ops.add(prefix, "cleanup")),
-                    Opt(:editable, :hstretch),
-                    # combo box label
-                    _("Cleanup algorithm"),
-                    cleanup_items(Ops.get_string(data, "cleanup", ""))
-                  )
-                )
-              ),
-              HSpacing(0.4)
-            )
-          ),
-          HSpacing()
-        )
+      args = {}
+      pre_args = {}
+      ret = nil
+
+      loop do
+        ret = UI.UserInput
+        args = get_modify_args(snapshot)
+        pre_args = get_modify_args(snapshot.pre) if snapshot.post?
+        break if ret == :ok || ret == :cancel
       end
 
-      if type != :POST
-        cont = VBox(
-          # popup label, %{num} is number
-          Label(_("Modify Snapshot %{num}") % { :num => num }),
-          snapshot_term.call("", snapshot)
-        )
-      else
-        cont = VBox(
-          # popup label, %{pre} and %{post} are numbers
-          Label(_("Modify Snapshot %{pre} and %{post}") % { :pre => pre_num, :post => num }),
-          # label
-          Left(Label(_("Pre (%{pre})") % { :pre => pre_num })),
-          snapshot_term.call("pre_", pre_snapshot),
-          VSpacing(),
-          # label
-          Left(Label(_("Post (%{post})") % { :post => num })),
-          snapshot_term.call("", snapshot)
-        )
+      UI.CloseDialog
+      if ret == :ok
+        modified = snapshot.update(args) if snapshot_modified(snapshot, args)
+
+        if snapshot.post? && snapshot_modified(snapshot.pre, pre_args)
+          modified = snapshot.pre.update(pre_args)
```

Comment:
good point, it should not be valid and in case the post snapshot is updated but pre isn't maybe we should do a rollback